cloud/filestore/libs/service_local/service.cpp

```diff
@@ -253,7 +267,81 @@ FILESTORE_SERVICE_LOCAL_ASYNC(FILESTORE_IMPLEMENT_METHOD_ASYNC)
     }
 
     template <typename T>
-    typename T::TResult Execute(typename T::TRequest& request)
+    void ProfileLogInit(
+        NProto::TProfileLogRequestInfo& logRequest,
+        const typename T::TRequest& request)
+    {
+        logRequest.SetRequestType(static_cast<ui32>(T::RequestType));
+        logRequest.SetTimestampMcs(Now().MicroSeconds());
+        InitProfileLogRequestInfo(logRequest, request);
+    }
+
+    template <typename T>
+    void ProfileLogFinilize(
+        NProto::TProfileLogRequestInfo&& logRequest,
+        const typename T::TResponse& response,
+        TString&& fsId)
+    {
+        FinalizeProfileLogRequestInfo(logRequest, response);
+        logRequest.SetDurationMcs(
+            Now().MicroSeconds() - logRequest.GetTimestampMcs());
+        logRequest.SetErrorCode(response.GetError().GetCode());
+        ProfileLog->Write({std::move(fsId), std::move(logRequest)});
+    }
+
+    template <typename T>
+    void ProfileLogFinilize(
+        NProto::TProfileLogRequestInfo&& logRequest,
+        TString&& fsId)
+    {
+        logRequest.SetDurationMcs(
+            Now().MicroSeconds() - logRequest.GetTimestampMcs());
+        ProfileLog->Write({std::move(fsId), std::move(logRequest)});
+    }
+
+    template <typename T>
+    typename T::TResult ExecuteWithPerfLog(typename T::TRequest& request)
+    {
+        NProto::TProfileLogRequestInfo logRequest;
+        ProfileLogInit<T>(logRequest, request);
+
+        auto response = Execute<T>(request, logRequest);
+
+        ProfileLogFinilize<T>(
+            std::move(logRequest),
+            response,
+            GetFileSystemId(request));
+
+        return response;
+    }
+
+    template <typename T>
+    typename T::TResult ExecuteWithPerfLogAsync(typename T::TRequest& request)
+    {
+        auto logRequest = std::make_shared<NProto::TProfileLogRequestInfo>();
+        ProfileLogInit<T>(*logRequest, request);
+
+        return Execute<T>(request, *logRequest)
+            .Subscribe(
+                [ptr = weak_from_this(),
+                 logRequest = std::move(logRequest),
+                 fsId = GetFileSystemId(request)](auto&) mutable
+                {
+                    auto self = ptr.lock();
+                    if (!self) {
+                        return;
+                    }
+
```

Comment:
I had to propagate `logRequest` to the Read/WriteAsync method so it will fill the response. Filling the response in subscribe handler won't work since if we call `future.GetValue()` it will break durable client that calls `future.ExtractValue()`